File: backend/app/core/monitoring.py
"""
Monitoring configuration with Sentry and structured logging.
"""
import logging
import sys
from typing import Optional

try:
    import sentry_sdk
    from sentry_sdk.integrations.fastapi import FastApiIntegration
    from sentry_sdk.integrations.sqlalchemy import SqlalchemyIntegration
    from sentry_sdk.integrations.redis import RedisIntegration
    from sentry_sdk.integrations.logging import LoggingIntegration
    SENTRY_AVAILABLE = True
except ImportError:
    SENTRY_AVAILABLE = False

try:
    import structlog
    STRUCTLOG_AVAILABLE = True
except ImportError:
    STRUCTLOG_AVAILABLE = False

from app.core.config import settings


def setup_sentry() -> None:
    """Initialize Sentry for error tracking."""
    if not SENTRY_AVAILABLE:
        logging.warning("Sentry SDK not installed. Install with: pip install sentry-sdk")
        return
    
    if not settings.sentry_dsn:
        logging.info("Sentry DSN not configured. Skipping Sentry initialization.")
        return
    
    sentry_sdk.init(
        dsn=settings.sentry_dsn,
        environment=settings.sentry_environment,
        release=f"flowto-backend@{settings.app_version}",
        traces_sample_rate=0.1,  # 10% of transactions for performance monitoring
        profiles_sample_rate=0.1,  # 10% for profiling
        integrations=[
            FastApiIntegration(transaction_style="endpoint"),
            SqlalchemyIntegration(),
            RedisIntegration(),
            LoggingIntegration(
                level=logging.INFO,  # Capture info and above as breadcrumbs
                event_level=logging.ERROR  # Send errors as events
            ),
        ],
        # Filter sensitive data
        before_send=before_send_sentry,
        # Performance monitoring
        enable_tracing=True,
    )
    
    logging.info(f"Sentry initialized for environment: {settings.sentry_environment}")


def before_send_sentry(event, hint):
    """
    Filter sensitive data before sending to Sentry.
    Remove passwords, tokens, API keys, etc.
    """
    # Remove sensitive headers
    if "request" in event and "headers" in event["request"]:
        headers = event["request"]["headers"]
        sensitive_headers = ["authorization", "cookie", "x-api-key"]
        for header in sensitive_headers:
            if header in headers:
                headers[header] = "[Filtered]"
    
    # Remove sensitive query parameters
    if "request" in event and "query_string" in event["request"]:
        query = event["request"]["query_string"]
        if "password" in query or "token" in query:
            event["request"]["query_string"] = "[Filtered]"
    
    # Remove sensitive data from extra context
    if "extra" in event:
        sensitive_keys = ["password", "token", "api_key", "secret", "iban"]
        for key in sensitive_keys:
            if key in event["extra"]:
                event["extra"][key] = "[Filtered]"
    
    return event


def setup_structured_logging() -> None:
    """
    Configure structured logging with structlog (JSON format).
    Falls back to standard logging if structlog not available.
    """
    if STRUCTLOG_AVAILABLE:
        structlog.configure(
            processors=[
                structlog.stdlib.filter_by_level,
                structlog.stdlib.add_logger_name,
                structlog.stdlib.add_log_level,
                structlog.stdlib.PositionalArgumentsFormatter(),
                structlog.processors.TimeStamper(fmt="iso"),
                structlog.processors.StackInfoRenderer(),
                structlog.processors.format_exc_info,
                structlog.processors.UnicodeDecoder(),
                structlog.processors.JSONRenderer()  # JSON output
            ],
            context_class=dict,
            logger_factory=structlog.stdlib.LoggerFactory(),
            cache_logger_on_first_use=True,
        )
        logging.info("Structured logging (structlog) initialized with JSON format")
    else:
        # Fallback to standard logging with JSON-like format
        logging.basicConfig(
            level=logging.INFO if settings.sentry_environment == "production" else logging.DEBUG,
            format='{"timestamp": "%(asctime)s", "level": "%(levelname)s", "logger": "%(name)s", "message": "%(message)s"}',
            handlers=[
                logging.StreamHandler(sys.stdout)
            ]
        )
        logging.warning("structlog not installed. Using standard logging. Install with: pip install structlog")


def get_logger(name: str):
    """
    Get a logger instance (structlog or standard).
    
    Args:
        name: Logger name (usually __name__)
        
    Returns:
        Logger instance
    """
    if STRUCTLOG_AVAILABLE:
        return structlog.get_logger(name)
    else:
        return logging.getLogger(name)


def log_request(
    method: str,
    path: str,
    status_code: int,
    duration_ms: float,
    user_id: Optional[str] = None
) -> None:
    """
    Log HTTP request with structured data.
    
    Args:
        method: HTTP method (GET, POST, etc.)
        path: Request path
        status_code: Response status code
        duration_ms: Request duration in milliseconds
        user_id: Optional user ID
    """
    logger = get_logger("flowto.api")
    
    log_data = {
        "event": "http_request",
        "method": method,
        "path": path,
        "status_code": status_code,
        "duration_ms": duration_ms,
    }
    
    if user_id:
        log_data["user_id"] = user_id
    
    if STRUCTLOG_AVAILABLE:
        logger.info("HTTP request", **log_data)
    else:
        logger.info(f"HTTP request: {log_data}")


def log_error(
    error: Exception,
    context: dict,
    user_id: Optional[str] = None
) -> None:
    """
    Log error with context.
    
    Args:
        error: Exception instance
        context: Additional context (endpoint, operation, etc.)
        user_id: Optional user ID
    """
    logger = get_logger("flowto.errors")
    
    log_data = {
        "event": "error",
        "error_type": type(error).__name__,
        "error_message": str(error),
        **context
    }
    
    if user_id:
        log_data["user_id"] = user_id
    
    if STRUCTLOG_AVAILABLE:
        logger.error("Application error", **log_data, exc_info=True)
    else:
        logger.error(f"Application error: {log_data}", exc_info=True)

File: frontend/src/lib/monitoring.ts
/**
 * Frontend monitoring with Sentry.
 * Captures errors and performance metrics.
 */

declare global {
  interface Window {
    Sentry?: any;
  }
}

/**
 * Initialize Sentry for error tracking.
 * Call this early in your app initialization.
 */
export function initSentry() {
  const sentryDsn = process.env.NEXT_PUBLIC_SENTRY_DSN;
  const environment = process.env.NEXT_PUBLIC_ENVIRONMENT || "development";
  const version = process.env.NEXT_PUBLIC_VERSION || "0.1.0";

  if (!sentryDsn) {
    console.info("Sentry DSN not configured. Skipping Sentry initialization.");
    return;
  }

  // Check if Sentry SDK is available
  if (typeof window === "undefined" || !window.Sentry) {
    console.warn(
      "Sentry SDK not loaded. Install with: npm install @sentry/nextjs"
    );
    return;
  }

  try {
    window.Sentry.init({
      dsn: sentryDsn,
      environment,
      release: `flowto-frontend@${version}`,

      // Performance monitoring
      tracesSampleRate: 0.1, // 10% of transactions
      replaysSessionSampleRate: 0.1, // 10% of sessions
      replaysOnErrorSampleRate: 1.0, // 100% of sessions with errors

      // Integrations
      integrations: [
        new window.Sentry.BrowserTracing({
          // Trace HTTP requests
          tracePropagationTargets: [
            "localhost",
            /^https:\/\/api\.flowto\.fr/,
          ],
        }),
        new window.Sentry.Replay({
          maskAllText: true, // Mask all text for privacy
          blockAllMedia: true, // Block all media for privacy
        }),
      ],

      // Filter sensitive data
      beforeSend(event, hint) {
        // Remove sensitive data from breadcrumbs
        if (event.breadcrumbs) {
          event.breadcrumbs = event.breadcrumbs.map((breadcrumb) => {
            if (breadcrumb.data) {
              // Remove passwords, tokens, etc.
              const sensitiveKeys = [
                "password",
                "token",
                "apiKey",
                "secret",
                "authorization",
              ];
              sensitiveKeys.forEach((key) => {
                if (key in breadcrumb.data!) {
                  breadcrumb.data![key] = "[Filtered]";
                }
              });
            }
            return breadcrumb;
          });
        }

        // Remove sensitive data from extra context
        if (event.extra) {
          const sensitiveKeys = ["password", "token", "apiKey", "secret"];
          sensitiveKeys.forEach((key) => {
            if (key in event.extra!) {
              event.extra![key] = "[Filtered]";
            }
          });
        }

        return event;
      },
    });

    console.info(`Sentry initialized for environment: ${environment}`);
  } catch (error) {
    console.error("Failed to initialize Sentry:", error);
  }
}

/**
 * Capture an exception in Sentry.
 */
export function captureException(error: Error, context?: Record<string, any>) {
  if (typeof window !== "undefined" && window.Sentry) {
    if (context) {
      window.Sentry.setContext("error_context", context);
    }
    window.Sentry.captureException(error);
  } else {
    console.error("Sentry not initialized:", error);
  }
}

/**
 * Set user context in Sentry.
 */
export function setUser(user: { id: string; email?: string } | null) {
  if (typeof window !== "undefined" && window.Sentry) {
    if (user) {
      window.Sentry.setUser({
        id: user.id,
        email: user.email,
      });
    } else {
      window.Sentry.setUser(null);
    }
  }
}

/**
 * Add breadcrumb for tracking user actions.
 */
export function addBreadcrumb(
  message: string,
  category: string,
  level: "info" | "warning" | "error" = "info",
  data?: Record<string, any>
) {
  if (typeof window !== "undefined" && window.Sentry) {
    window.Sentry.addBreadcrumb({
      message,
      category,
      level,
      data,
    });
  }
}

